File: DynamicProgramming/27/kprobBottomUp.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number} row
 * @param {number} column
 * @return {number}
 */

const directions = [
  [-2, -1],
  [-2, 1],
  [-1, 2],
  [1, 2],
  [2, -1],
  [2, 1],
  [-1, -2],
  [1, -2],
];

var knightProbability = function (n, k, row, column) {
  if (row < 0 || row >= n || column < 0 || column >= n || n == 0) return 0;
  if (k === 0) return 1;
  const grid = new Array(k + 1)
    .fill(0)
    .map(() => new Array(n).fill(0).map(() => new Array(n).fill(0)));
  grid[0][row][column] = 1;
  for (let itr = 0; itr < k; itr++) {
    for (let r = 0; r < n; r++) {
      for (let c = 0; c < n; c++) {
        if (grid[itr][r][c] > 0) {
          for (const direction of directions) {
            const newRow = r + direction[0];
            const newCol = c + direction[1];
            if (newRow < 0 || newRow >= n || newCol < 0 || newCol >= n) {
              continue;
            }
            grid[itr + 1][newRow][newCol] =
              grid[itr + 1][newRow][newCol] + (grid[itr][r][c] / 8);
          }
        }
      }
    }
  }
  let sum = 0;
  for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
      sum += grid[k][r][c];
    }
  }
  return sum;
};

File: DynamicProgramming/27/kprob_bottom_up.py
class Solution:
    def knightProbability(self, n: int, k: int, row: int, column: int) -> float:
        if row < 0 or row >= n or column < 0 or column >=n or n == 0:
            return 0
        if k == 0: 
            return 1
        directions = [
            [-2, -1],
            [-2, 1],
            [-1, 2],
            [1, 2],
            [2, -1],
            [2, 1],
            [-1, -2],
            [1, -2]
        ]
        grid = [[[0]*n for _ in range(n)] for _ in range(k+1)]
        grid[0][row][column] = 1

        for itr in range(k):
            for r in range(n):
                for c in range(n):
                    if grid[itr][r][c] > 0:
                        for direction in directions:
                            newRow = r + direction[0]
                            newCol = c + direction[1]
                            if newRow < 0 or newRow >= n or newCol < 0 or newCol >= n:
                                continue
                            grid[itr+1][newRow][newCol] = grid[itr+1][newRow][newCol] + (grid[itr][r][c]/8)
        
        sum = 0
        for i in range(n):
            for j in range(n):
                sum += grid[k][i][j]
        return sum

